File: problems_2/7.py
# Problem 7: Single Number (Find Unique)
# **Difficulty: ** Easy
# **Time:** 10 minutes
# **Pattern:** XOR properties
#
# Find the element that appears only once while all others appear twice.


def single_number(nums):
    """
    Find the single element (all others appear twice).

    Examples:
    Input: [2, 2, 1]
    Output: 1

    Input: [4, 1, 2, 1, 2]
    Output: 4

    Use O(1) space complexity.
    """
    # Your code here
    nums = sorted(nums)
    print(nums)
    if len(nums) == 1:
        return nums[0]
    for n in range(0, len(nums) - 1, 2):
        print(f"a: {nums[n]} b: {nums[n + 1]}")
        if nums[n] != nums[n + 1]:
            return nums[n]
    return nums[-1]


# Test cases
assert single_number([2, 2, 1]) == 1
assert single_number([4, 1, 2, 1, 2]) == 4
assert single_number([1]) == 1
assert single_number([3, 3, 7, 7, 10, 11, 11]) == 10
print("All tests passed!")

File: cpp/problems_1/7.cpp
// #### Problem 7: Best Time to Buy and Sell Stock
// **Difficulty:** Easy
// **Time:** 15 minutes
// **Pattern:** Single pass with min tracking
//
// You are given an array `prices` where `prices[i]` is the price of a given
// stock on the `i`th day. Find the maximum profit you can achieve.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

int maxProfit(std::vector<int> &prices) {
  /*
  Find maximum profit from buying and selling stock once.

  Examples:
  Input: prices = [7,1,5,3,6,4]
  Output: 5 (buy at 1, sell at 6)

  Input: prices = [7,6,4,3,1]
  Output: 0 (no profit possible)
  */
  int max_value = 0;
  // Your code here
  for (size_t i = 0; i < prices.size(); i++) {
    for (size_t j = 1; j < prices.size() - i; j++) {
      if (prices[i] - prices[i + j] < max_value &&
          prices[i] - prices[i + j] < 0) {
        max_value = prices[i] - prices[i + j];
      }
    }
  }
  return abs(max_value);
}

int main() {
  std::vector<int> prices1 = {7, 1, 5, 3, 6, 4};
  assert(maxProfit(prices1) == 5);

  std::vector<int> prices2 = {7, 6, 4, 3, 1};
  assert(maxProfit(prices2) == 0);

  std::vector<int> prices3 = {1, 2, 3, 4, 5};
  assert(maxProfit(prices3) == 4);

  std::cout << "Problem 7: All tests passed!" << std::endl;
  return 0;
}
